compiler/codegen/CCData.hpp

```diff
@@ -0,0 +1,207 @@
+       *
+       * @param[In] storage A pointer to a memory buffer that CCData will use.
+       * @param[In] sizeBytes The amount of data (in bytes) that the buffer contains.
+       */
+      CCData(uint8_t * const storage, const size_t sizeBytes);
+
+      ~CCData();
+
+      /**
+       * @brief Puts the given value in the table, optionally mapped to the given key (if any), aligned to the value's natural type, and returns the index to the value. Synchronized.
+
+       * @tparam T The type of the value to put in the table. The type must be TriviallyCopyable. A static assertion checks for the std::is_trivially_copyable<T> type trait on capable compilers.
+       * @param[In] value The value to put.
+       * @param[In] key Optional. The key to map the resulting index to. Without a key the index is the only reference to the data.
+       * @param[Out] index The index that refers to the value.
+       */
+      template <typename T>
+      bool put(const T value, const key_t * const key, index_t &index);
+
+      /**
+       * @brief Puts the given value in the table, optionally mapped to the given key (if any), aligned to the given boundary (in bytes), and returns the index to the value. Synchronized.
+
+       * @param[In] value Optional. A pointer to the value to put. If null, no data will be copied but the space will be allocated none the less.
+       * @param[In] sizeBytes The size of the value pointed to.
+       * @param[In] alignmentBytes The alignment (in bytes) to align the value to.
+       * @param[In] key Optional. The key to map the resulting index to. Without a key the index is the only reference to the data. If the key is already mapped to an index the operation will return the index and true, but no data will be written.
+       * @param[Out] index The index that refers to the value.
+       * @return True if the value was placed in the table, or the key was already mapped to an index, false otherwise.
+       */
+      bool put(const uint8_t * const value, const size_t sizeBytes, const size_t alignmentBytes, const key_t * const key, index_t &index);
+
+      /**
+       * @brief Gets the value referred to by the index from the table.
+
+       * @param[In] T The type of the value to get from the table. The type must be TriviallyCopyable. A static assertion checks for the std::is_trivially_copyable<T> type trait on capable compilers.
+       * @param[In] index The index that refers to the value.
+       * @param[out] value A reference to the value to write the result to.
+       * @return True if the value was placed in the table, false otherwise.
+       */
+      template <typename T>
+      bool get(const index_t index, T &value) const;
+
+      /**
+       * @brief Gets the value referred to by the index from the table.
+
+       * @param[In] index The index that refers to the value.
+       * @param[in] value A pointer to the value to write the result to. This parameter is ignored unless this function returns true.
```

Comment:
Should this be an `[Out]`?

I think a better way of phrasing this would be "This parameter is not changed unless the operations succeeds and this function returns true".

Reply:
Yes it should be `[Out]`, noted.